engine/executor/dm/worker.go

```diff
@@ -0,0 +1,242 @@
+		return err
+	}
+	if err := w.tryUpdateStatus(ctx); err != nil {
+		return err
+	}
+	return w.messageAgent.Tick(ctx)
+}
+
+// Workload implements lib.WorkerImpl.Worload
+func (w *dmWorker) Workload() model.RescUnit {
+	log.L().Info("dmworker.Workload")
+	return 0
+}
+
+// OnMasterFailover implements lib.WorkerImpl.OnMasterFailover
+func (w *dmWorker) OnMasterFailover(reason lib.MasterFailoverReason) error {
+	log.L().Info("dmworker.OnMasterFailover")
+	return nil
+}
+
+// OnMasterMessage implements lib.WorkerImpl.OnMasterMessage
+func (w *dmWorker) OnMasterMessage(topic p2p.Topic, message p2p.MessageValue) error {
+	log.L().Info("dmworker.OnMasterMessage", zap.String("topic", topic), zap.Any("message", message))
+	return nil
+}
+
+// CloseImpl implements lib.WorkerImpl.CloseImpl
+func (w *dmWorker) CloseImpl(ctx context.Context) error {
+	// unreigster jobmaster sender
+	if err := w.messageAgent.UpdateSender(w.masterID, nil); err != nil {
+		return err
+	}
+	w.unitHolder.Close(ctx)
+	w.messageAgent.Close(ctx)
+	return nil
+}
+
+// closeAndExit closes the task and exits.
+func (w *dmWorker) closeAndExit(ctx context.Context, status libModel.WorkerStatus) error {
+	err := w.CloseImpl(ctx)
+	if err != nil {
+		log.L().Warn("fail to close task", log.ShortError(err))
+	}
+	return w.Exit(ctx, status, nil)
+}
+
+// setupstorage opens and configs external storage
+func (w *dmWorker) setupstorage(ctx context.Context) error {
+	rid := dm.NewDMResourceID(w.cfg.Name, w.cfg.SourceID)
+	h, err := w.OpenStorage(ctx, rid)
+	for status.Code(err) == codes.Unavailable {
+		log.L().Info("simple retry", zap.Error(err))
+		time.Sleep(time.Second)
+		h, err = w.OpenStorage(ctx, rid)
+	}
+	if err != nil {
+		return errors.Trace(err)
+	}
+	w.storageWriteHandle = h
+	w.cfg.ExtStorage = h.BrExternalStorage()
+	return nil
+}
+
+// persistStorage persists storage.
+func (w *dmWorker) persistStorage(ctx context.Context) error {
+	return w.storageWriteHandle.Persist(ctx)
+}
+
+// tryUpdateStatus updates status when task stage changed.
+func (w *dmWorker) tryUpdateStatus(ctx context.Context) error {
+	stage := w.unitHolder.Stage()
+	if stage == w.getStage() {
+		return nil
+	}
+	log.L().Info("task stage changed", zap.Int("from", int(stage)), zap.Int("to", int(w.stage)))
+	w.setStage(stage)
+
+	status := w.workerStatus()
+	if stage != metadata.StageFinished {
+		return w.UpdateStatus(ctx, status)
+	}
+
+	if w.workerType == lib.WorkerDMDump {
+		if err := w.persistStorage(ctx); err != nil {
+			log.L().Error("failed to persist storage", zap.Error(err))
+			return nil
+		}
+	}
+	return w.closeAndExit(ctx, status)
+}
+
+// workerStatus gets worker status.
+func (w *dmWorker) workerStatus() libModel.WorkerStatus {
+	stage := w.getStage()
+	code := libModel.WorkerStatusNormal
+	if stage == metadata.StageFinished {
+		code = libModel.WorkerStatusFinished
+	}
+	status := runtime.TaskStatus{Unit: w.workerType, Task: w.taskID, Stage: stage}
```

Comment:
seems we will get a Normal worker whose stage is Paused. Maybe we can list the combinations of worker statsu and task stage in future.